src/telliot_feeds/sources/gas_price_oracle.py

```diff
@@ -2,127 +2,125 @@
-                128: "ht",
-                1285: "movr",
-                122: "fuse",
-            }
-
-            url = (
-                f"https://owlracle.info/"
-                f"{networks[self.chainId]}"
-                "/history?"
-                f"from={int(self.timestamp)}"
-                f"&to={int(self.timestamp) + 100}"
-            )
-
-            with requests.Session() as s:
-                s.mount("https://", adapter)
-                try:
-                    return s.get(url=url, timeout=0.5, headers={"User-Agent": "Custom"})
-
-                except requests.exceptions.RequestException as e:
-                    logger.error(f"GasPriceOracle API error: {e}")
-                    return None
-
-                except requests.exceptions.Timeout as e:
-                    logger.error(f"GasPriceOracle API timed out: {e}")
-                    return None
+    def binary_search_block(self) -> Optional[BlockData]:
+        """Binary search for the block closest to the target timestamp (not later)
 
+        Returns:
+            The block closest to the target timestamp (not later)
+        """
+        if not self.web3:
+            raise ValueError("Web3 not instantiated")
+        if not self.timestamp:
+            raise ValueError("Timestamp not provided")
+
+        right_block_number: int = self.web3.eth.block_number
+        left_block_number = 0
+        target_timestamp = self.timestamp
+
+        while left_block_number <= right_block_number:
+            mid = (left_block_number + right_block_number) // 2
+            try:
+                mid_block = self.web3.eth.get_block(mid)
+            except ExtraDataLengthError as e:
+                logger.info(f"POA chain detected. Injecting POA middleware in response to exception: {e}")
+                try:
+                    self.web3.middleware_onion.inject(geth_poa_middleware, layer=0)
+                except ValueError as e:
+                    logger.error(f"Unable to inject web3 middleware for POA chain connection: {e}")
+                mid_block = self.web3.eth.get_block(mid)
+            except Exception as e:
+                logger.error(f"Failed to fetch block info: {e}")
+                return None
+
+            mid_block_timestamp = mid_block["timestamp"]
+
+            # Check for exact match
+            if mid_block_timestamp == target_timestamp:
+                return mid_block
+
+            if mid_block_timestamp < target_timestamp:
+                left_block_number = mid + 1
+            else:
+                right_block_number = mid - 1
+
+        # If we've exited the loop, that means the target timestamp is not equal to any block timestamp,
+        # we'll find the block that is closest the target timestamp (not later).
+        left_block = self.web3.eth.get_block(left_block_number)
+        right_block = self.web3.eth.get_block(right_block_number)
+        if left_block["timestamp"] > target_timestamp:
```

Comment:
What if left block timestamp and right block timestamp are both greater than target timestamp? This will return left block in that case. Is this the intended behavior?

Reply:
I don't think that would happen where both are greater than the target timestamp since this is handled in the search. I'll revisit this